crates/ark/src/interface.rs

```diff
@@ -243,6 +244,16 @@ pub struct RMain {
     /// `WriteConsole` output diverted from IOPub is stored here. This is only used
     /// to return R output to the debugger.
     pub(crate) captured_output: String,
+
+    /// Whether we should preserve focus when stopping in a debug session. We
+    /// should only preserve focus if we're explicitly stepping through code as
+    /// opposed to evaluating an expression in the debugger console.
+    /// See https://github.com/posit-dev/positron/issues/3151.
+    debug_preserve_focus: bool,
+
+    /// The stack of frames we saw the last time we stopped. Used as a mostly
+    /// reliable indication of whether we moved since last time.
+    debug_last_stack: Vec<FrameInfoId>,
 }
```

Comment:
`RMainDap` already contains many "context related" pieces of information, like 

```rust
    /// The current call emitted by R as `debug: <call-text>`.
    call_text: DebugCallText,

    /// The last known `start_line` for the active context frame.
    last_start_line: Option<i64>,
```

I think I would really prefer that these move there as simply `preserve_focus` and `last_stack`.

I have a feeling it will clean some state management up too, as `self.debug_last_stack = vec![];` can move into `stop_debug()` where we already do cleanup like that.

Reply:
Actually it's possible you might be able to move them all the way down to `Dap` itself!

That actually feels like the right spot, and it's possible you won't even need to retain `last_stack` if you do that because in `Dap`'s `start_debug()` it looks like you can just compare the new `stack` to the old `self.stack` to finalize the `preserve_focus` value.